File: apps/ios/NOVA/Models/Conversation.swift
import Foundation

/// A thread of messages between the owner and NOVA.
struct Conversation: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    /// Derived server-side from the first message rather than asked for.
    let title: String?
    let messageCount: Int
    let lastMessageAt: Date?
    let createdAt: Date

    var displayTitle: String {
        title ?? "New conversation"
    }
}

/// A conversation with its messages, oldest first.
struct ConversationDetail: Codable, Identifiable, Equatable, Sendable {
    let id: UUID
    let title: String?
    let messageCount: Int
    let lastMessageAt: Date?
    let createdAt: Date
    let messages: [ChatMessage]
}

/// One turn.
struct ChatMessage: Codable, Identifiable, Equatable, Sendable {
    enum Role: String, Codable, Sendable {
        case user, assistant
    }

    let id: UUID
    let role: Role
    let content: String
    let createdAt: Date
    let model: String?
    let latencyMs: Int?

    var isFromNova: Bool { role == .assistant }
}

extension ChatMessage {
    /// A message shown before the server has confirmed it.
    ///
    /// Given a client-side id so SwiftUI can animate it in immediately; the
    /// server echoes the stored turn as the stream's first event and the
    /// optimistic copy is replaced by it.
    static func pending(_ content: String) -> ChatMessage {
        ChatMessage(
            id: UUID(),
            role: .user,
            content: content,
            createdAt: .now,
            model: nil,
            latencyMs: nil
        )
    }

    /// The assistant turn being assembled from stream deltas.
    static func streaming(_ text: String, id: UUID) -> ChatMessage {
        ChatMessage(
            id: id,
            role: .assistant,
            content: text,
            createdAt: .now,
            model: nil,
            latencyMs: nil
        )
    }
}

// MARK: - Requests

struct CreateConversationRequest: Encodable, Sendable {
    let title: String?
}

struct SendMessageRequest: Encodable, Sendable {
    let content: String
}

/// Both halves of a completed non-streaming exchange.
struct MessageExchange: Decodable, Sendable {
    let userMessage: ChatMessage
    let assistantMessage: ChatMessage
}

File: services/api/src/nova/schemas/conversation.py
"""Conversation payloads."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

# Long enough for a paragraph dictated by voice, short enough that a single
# turn cannot blow out the context window or the bill.
MAX_MESSAGE_LENGTH = 4000


class MessageRead(BaseModel):
    """One stored turn."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    role: Literal["user", "assistant"]
    content: str
    created_at: datetime
    model: str | None = None
    latency_ms: int | None = None


class ConversationRead(BaseModel):
    """A thread, without its messages."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    title: str | None
    message_count: int
    last_message_at: datetime | None
    created_at: datetime


class ConversationDetail(ConversationRead):
    """A thread with its newest messages, oldest first.

    ``messages`` is capped at the newest 200; ``message_count`` is the real
    total, so ``len(messages) < message_count`` means older ones were left out.
    """

    messages: list[MessageRead] = Field(default_factory=list)


class CreateConversationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    title: str | None = Field(default=None, max_length=120)


class SendMessageRequest(BaseModel):
    """Something said to NOVA."""

    model_config = ConfigDict(extra="forbid")

    content: str = Field(min_length=1, max_length=MAX_MESSAGE_LENGTH)

    @field_validator("content")
    @classmethod
    def _reject_blank(cls, value: str) -> str:
        stripped = value.strip()
        if not stripped:
            raise ValueError("Message must not be blank.")
        return stripped


class MessageExchange(BaseModel):
    """Both halves of a completed non-streaming exchange."""

    user_message: MessageRead
    assistant_message: MessageRead
